### PR TITLE
Add URL Configuration support to azurerm_application_gateway

### LINKED ISSUE #7565: Support for URL rewrite in azurerm_application_gateway
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Microsoft has a preview feature which allows URL rewrites to be part of a rewrite rule set. 

### New or Affected Resource(s)

* azurerm_applicatiion_gateway

### Potential Terraform Configuration

```hcl
  rewrite_rule_set {
    name = "example_rewrite_rule_set"

    rewrite_rule {
      name          = "example_rule"
      rule_sequence = 100
      
      condition {
        ignore_case = true
        negate      = true
        pattern     = "\\/\\w+"
        variable    = "var_request_uri"
      }
      url_configuration {
        components     = url_path
        url_path_value  = "/example.aspx"
      }
    }
}
```

### PR DESCRIPTION
**Pull request by dhensby**
Fixes #7565

This is my first pass at trying to get `URLConfiguration` support into terraform (see https://docs.microsoft.com/en-us/rest/api/application-gateway/applicationgateways/get#applicationgatewayurlconfiguration).

This will allow the rewriting of URLs in application gateway.

To-dos:

- [x] Should we validate that at least one of `url_configuration.url_path` or `url_configuration.url_query` is supplied?
- [x] Update docs
- [x] tests??

---

example usage:

```hcl

resource "azurerm_application_gateway" "gateway" {
  ...
  rewrite_rule_set {
    name = local.rewrite_rule_set_name

    rewrite_rule {
      name          = local.rewrite_rule_name
      rule_sequence = 1

      condition {
        variable = "var_uri_path"
        pattern  = ".*article/(.*)/(.*)"
      }

      url {
        path         = "/article.aspx"
        query_string = "id={var_uri_path_1}&title={var_uri_path_2}"
      }
    }
  }
  ...
}

```

Test data based on example here https://docs.microsoft.com/en-us/azure/application-gateway/rewrite-url-portal

### CODE REVIEW

**Comment by dhensby** [comment]
I have compiled and tested this locally and successfully deployed a url_rewrite against an app gateway.

**Comment by jackofallops** [comment]
Hi @dhensby - Could you revert any changes to `CHANGELOG.md` - This file is updated by maintainers when we actually merge and should not be part of a PR content.

Thanks!

File: azurerm/internal/services/network/application_gateway_resource.go
Comment (katbyte) [inline azurerm/internal/services/network/application_gateway_resource.go:1005]: could we validate this? or are empty strings valid?

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks for the PR @dhensby - overall looks great, but i have some questions i've left inline to address before merge

File: azurerm/internal/services/network/application_gateway_resource.go
Comment (dhensby) [reply azurerm/internal/services/network/application_gateway_resource.go:992]: I'm happy to change this, but I've kept with `url_configuration` because that [more closely matches the API exposed by azure](https://docs.microsoft.com/en-us/rest/api/application-gateway/applicationgateways/createorupdate#applicationgatewayrewriteruleactionset) and this also matches with the other options (response_header_configuration, request_header_configuration).

Let me know if you think we should match the underlying API or go with a more concise name.

File: website/docs/r/application_gateway.html.markdown
Comment (dhensby) [reply website/docs/r/application_gateway.html.markdown:585]: I think this is a good idea; however I'm not entirely sure of the best way to do this? Any pointers would be greatly received (examples of other resources validating this kind either/or requirement 😇

File: azurerm/internal/services/network/application_gateway_resource.go
Comment (dhensby) [reply azurerm/internal/services/network/application_gateway_resource.go:1005]: Interestingly, empty strings are allowed; I've just tried it in the portal and it's accepted.

2021-03-18_18-03-23_97d8af6d-fee4-4226-9e97-7ffbd8966a52

File: azurerm/internal/services/network/application_gateway_resource.go
Comment (dhensby) [reply azurerm/internal/services/network/application_gateway_resource.go:1001]: Empty string is an allowed value

File: azurerm/internal/services/network/application_gateway_resource.go
Comment (dhensby) [reply azurerm/internal/services/network/application_gateway_resource.go:992]: @katbyte - what were you final thoughts on this?

Comment (katbyte) [reply azurerm/internal/services/network/application_gateway_resource.go:992]: We don't follow the underlying api in a number of places either for ux/clarity/its wrong so i think its best to change this for a cleaner config in tf

File: azurerm/internal/services/network/application_gateway_resource.go
Comment (katbyte) [reply azurerm/internal/services/network/application_gateway_resource.go:1005]: 👍

File: website/docs/r/application_gateway.html.markdown
Comment (katbyte) [reply website/docs/r/application_gateway.html.markdown:585]: in the create function we can just check that either path or query string have a value and if nether is set return an error that'll bubble up

File: azurerm/internal/services/network/application_gateway_resource.go
Comment (dhensby) [reply azurerm/internal/services/network/application_gateway_resource.go:992]: OK - I also had a think about this and I thought that `url` is misleading because one would expect that to literally be `url = "https://example.com", but this isn't like that.

I'll make the change and we can change it again if we think there's a better name for it

Comment (dhensby) [reply azurerm/internal/services/network/application_gateway_resource.go]: I've made the change - I did notice when I did it that the `_configuration` string is added to quite a few of the setting names for the `application_gateway_resource`, so something about keeping it here feels more natural to me

File: website/docs/r/application_gateway.html.markdown
Comment (dhensby) [reply website/docs/r/application_gateway.html.markdown:585]: I've added a check and return an error when neither are set

File: azurerm/internal/services/network/application_gateway_resource.go
Comment (katbyte) [reply azurerm/internal/services/network/application_gateway_resource.go]: those will likely be renamed in the future, it's just been a low priority to rename things

File: CHANGELOG.md
Comment (katbyte) [inline CHANGELOG.md:72]:
```suggestion
ENHANCEMENTS:

* `azurerm_application_gateway.rewrite_rule` add support for URL rewriting with `url_configuration` block ([#10950](https://github.com/terraform-providers/terraform-provider-azurerm/pull/10950))

```

we do this by hand

**Review by katbyte — APPROVED** [review]
Thanks @dhensby - LGTM 👍

**Comment by ido-tr** [comment]
Thanks for this very impotent update,
I think there is a minor bug in the result (below)
when using 
url {
         path =foo
        }

the result is "Both URL path and URL query string" and not "URL path" (screenshot below)

image

**Comment by dhensby** [comment]
I think this is just how the front end in the portal is showing it.

That option isn't actually part of the rest API, it's just how the portal shows the interface.

**Comment by ido-tr** [comment]
@dhensby - after tests, the query strings are not working when it set to "Both URL path and URL query string" and not to "URL path" only.
In my tf module, the path is set to "path" only and not both "path" and "query_string"  (below)

url {
path =foo
}

**Comment by dhensby** [comment]
@ido-tr I don't really understand what you're saying.

I've gone through to double check this all works as expected and I don't see any issues:

### Both path and query_string set:

```hcl
  rewrite_rule_set {
    name = "test-rewrite-ruleset"

    rewrite_rule {
      name          = "test-rewrite-rule"
      rule_sequence = 1

      condition {
        variable = "var_uri_path"
        pattern  = ".*article/(.*)/(.*)"
      }

      url {
        path         = "/article.aspx"
        query_string = "id={var_uri_path_1}&title={var_uri_path_2}"
      }
    }
  }
```

In the portal:

2021-04-19_10-27-07_a3ba29b4-c668-4d3d-b12c-f2f22127cf90

### Just query_string:

```hcl
  rewrite_rule_set {
    name = "test-rewrite-ruleset"

    rewrite_rule {
      name          = "test-rewrite-rule"
      rule_sequence = 1

      condition {
        variable = "var_uri_path"
        pattern  = ".*article/(.*)/(.*)"
      }

      url {
        query_string = "id={var_uri_path_1}&title={var_uri_path_2}"
      }
    }
  }
```

In the portal:

2021-04-19_10-29-07_c4b041a8-848a-4790-aa00-a85df8e25b75


### Just path:

```hcl
  rewrite_rule_set {
    name = "test-rewrite-ruleset"

    rewrite_rule {
      name          = "test-rewrite-rule"
      rule_sequence = 1

      condition {
        variable = "var_uri_path"
        pattern  = ".*article/(.*)/(.*)"
      }

      url {
        path         = "/article.aspx"
      }
    }
  }
```

In the portal:

2021-04-19_10-33-25_0eb77c2f-bac0-4926-8b41-50c21e7ea8dc


All those are working as expected; can you provide me an example that doesn't work?

**Comment by dhensby** [comment]
Having inspected the state file, I think I see the problem. When no value is provided the state file shows that there is a value of `""` (an empty string), which probably means the `query_string` is essentially being replaced by the rewrite rule.

I'm not entirely sure why this is the case, but I'll do some investigation as to why empty strings are going through when no value is provided.

**Comment by dhensby** [comment]
OK - this is where my lack of Go/terraform knowledge is letting me down - I can't see why empty values are being passed when values are missing.

It appears it may not be possible to distinguish a missing value vs empty string for `TypeString`, in which case we have a bit of a problem...

@katbyte any thoughts on this?

**Comment by ido-tr** [comment]
@dhensby - this is exactly my point

**Pull request merged**
